File: Data Structures & Algorithms/products-of-array-discluding-self/submission-3.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n=nums.size();
        // vector<int>pref(n);
        // vector<int> suff(n);
        vector<int> result(n);
        result[0]=1;
        // suff[n-1]=1;
        for(int i=1;i<n;i++){
            result[i]=nums[i-1]*result[i-1];
        }
        // for(int i=n-1; i>=0;i--){
        //     suff[i]=nums[i+1]*suff[i-1];
        // }
        int postfix=1;
        for(int i=n-1; i>=0;i--){
            result[i]=result[i]*postfix;
            postfix*= nums[i];
        
        }
        return result;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.java
class Solution {
    public int search(int[] nums, int target) {
        int n=nums.length;
        int l=0; int r=n-1;
        while(l<r){
            int mid=l+(r-l)/2;
            if(nums[mid]>nums[r]){
                l=mid+1;
            }
            else {
                r=mid;
            }
        }
        int pivot= l;
        int result= binanrySearch(nums, target,0,pivot-1);
        if(result!=-1){
            return result;
        }
        return binanrySearch(nums, target, pivot, n-1);
    }
    public int binanrySearch(int[]nums,int target,int l,int r){
        while(l<=r){
            int mid= l+(r-l)/2;
            if(nums[mid]== target){
                return mid;
            }
            else if(nums[mid]>target){
                r=mid-1;
            }
            else {
                l=mid+1;
            }
        }
        return -1;
    }
}
